src/parcsr_mv/par_csr_matop_device.c

```diff
@@ -1035,22 +1035,155 @@ hypre_ParCSRMatrixGetRowDevice( hypre_ParCSRMatrix  *mat,
- * option == 0, drop all the entries that are smaller than tol
- * TODO more options
+/* Get element-wise tolerances based on row norms for ParCSRMatrix
+ * NOTE: Keep the diagonal, i.e. elmt_tol = 0.0 for diagonals
+ * Output vectors have size nnz:
+ *    elmt_tols_diag[j] = tol * (norm of row i) for j in [ A_diag_i[i] , A_diag_i[i+1] )
+ *    elmt_tols_offd[j] = tol * (norm of row i) for j in [ A_offd_i[i] , A_offd_i[i+1] )
+ * type == -1, infinity norm,
+ *         1, 1-norm
+ *         2, 2-norm
  */
+template<HYPRE_Int type>
+__global__ void
+hypre_ParCSRMatrixDropSmallEntriesDevice_getElmtTols( HYPRE_Int      nrows,
+                                                      HYPRE_Real     tol,
+                                                      HYPRE_Int     *A_diag_i,
+                                                      HYPRE_Int     *A_diag_j,
+                                                      HYPRE_Complex *A_diag_a,
+                                                      HYPRE_Int     *A_offd_i,
+                                                      HYPRE_Complex *A_offd_a,
+                                                      HYPRE_Real     *elmt_tols_diag,
+                                                      HYPRE_Real     *elmt_tols_offd)
+{
+   HYPRE_Int row_i = hypre_cuda_get_grid_warp_id<1,1>();
+
+   if (row_i >= nrows)
+   {
+      return;
+   }
+
+   HYPRE_Int lane = hypre_cuda_get_lane_id<1>();
+   HYPRE_Int p, q;
+
+   /* sum row norm over diag part */
+   if (lane < 2)
+   {
+      p = read_only_load(A_diag_i + row_i + lane);
+   }
+   q = __shfl_sync(HYPRE_WARP_FULL_MASK, p, 1);
+   p = __shfl_sync(HYPRE_WARP_FULL_MASK, p, 0);
+
+   HYPRE_Real row_norm_i = 0.0;
+
+   for (HYPRE_Int j = p + lane; j < q; j += HYPRE_WARP_SIZE)
+   {
+      HYPRE_Complex val = A_diag_a[j];
+
+      if (type == -1)
+      {
+         row_norm_i = hypre_max(row_norm_i, hypre_cabs(val));
+      }
+      else if (type == 1)
+      {
+         row_norm_i += hypre_cabs(val);
+      }
+      else if (type == 2)
+      {
+         row_norm_i += val * val;
+      }
+   }
+
+   /* sum row norm over offd part */
+   if (lane < 2)
+   {
+      p = read_only_load(A_offd_i + row_i + lane);
+   }
+   q = __shfl_sync(HYPRE_WARP_FULL_MASK, p, 1);
+   p = __shfl_sync(HYPRE_WARP_FULL_MASK, p, 0);
+
+   for (HYPRE_Int j = p + lane; j < q; j += HYPRE_WARP_SIZE)
+   {
+      HYPRE_Complex val = A_offd_a[j];
+
+      if (type == -1)
+      {
+         row_norm_i = hypre_max(row_norm_i, hypre_cabs(val));
+      }
+      else if (type == 1)
+      {
+         row_norm_i += hypre_cabs(val);
+      }
+      else if (type == 2)
+      {
+         row_norm_i += val * val;
+      }
+   }
+
+   /* allreduce to get the row norm on all threads */
+   row_norm_i = warp_allreduce_sum(row_norm_i);
+   if (type == 2)
+   {
+      row_norm_i = sqrt(row_norm_i);
+   }
+
+   /* set elmt_tols_diag */
+   if (lane < 2)
+   {
+      p = read_only_load(A_diag_i + row_i + lane);
```

Comment:
`p = read_only_load(A_diag_i + row_i + lane);` read twice? Maybe having `p1, q1, p2, q2` to save these?